### PR TITLE
test/dhtest.c: fix resource leak

### PR DESCRIPTION
**Pull request by mspncp**
Reported by Coverity Scan (CID 1439136)
[extended tests]

### CODE REVIEW

**Comment by mspncp** [comment]
dhtest failure is related

**Comment by mspncp** [comment]
[Travis failure](https://travis-ci.org/openssl/openssl/jobs/426395939) is unrelated: `MarkInfo objects are deprecated`.

File: test/dhtest.c [hunk 1]
```diff
@@ -62,11 +62,15 @@ static int dh_test(void)
         || !TEST_true(DH_set0_pqg(dh, p, q, g)))
         goto err;
 
+    p1=p, p=NULL;
+    q1=q, q=NULL;
+    g1=g, g=NULL;
```
Comment (romen) [inline test/dhtest.c]: Does this kind of expression adhere to the code style guidelines?

Comment (romen) [reply test/dhtest.c]: (applies to the above 3 lines)

File: test/dhtest.c [hunk 2]
```diff
@@ -79,11 +83,12 @@ static int dh_test(void)
     if (!TEST_true(BN_set_word(priv_key, 1234L))
         || !TEST_true(DH_set0_key(dh, NULL, priv_key)))
         goto err;
+    priv_key1 = priv_key, priv_key = NULL;
```
Comment (romen) [inline test/dhtest.c]: Does this kind of expression adhere to the code style guidelines?

File: test/dhtest.c (hunk 1 above)
Comment (romen) [reply test/dhtest.c]: I'd also like if there was a comment pointing at `DH_set0_pgq` (and the related `BN_free`s in case of early error) as the motivation of `p = q = g = NULL`, while we want to save the addresses for the subsequent tests.

Comment (romen) [reply test/dhtest.c]: As an alternative the problem with `p`, `q` and `g` could be solved by replacing the first five `goto err`  with `goto early_err` and then adding at the end of the function:

```diff
     return ret;
+
+ early_err:
+    DH_free(dh);
+    BN_free(p);
+    BN_free(q);
+    BN_free(g);
+    BN_free(priv_key);
+    return 0;
 }
```
and omitting `BN_free()` on `p`, `q`, `g` and `priv_key` in the *normal* `err` handling.

This should allow to avoid `p1`, `q1`, `g1` and `priv_key1` at all.

File: test/dhtest.c [hunk 3]
```diff
@@ -62,11 +62,15 @@ static int dh_test(void)
         || !TEST_true(DH_set0_pqg(dh, p, q, g)))
         goto err;
 
+    p1=p, p=NULL;
```
Comment (kaduk) [inline test/dhtest.c]: I think I'm horribly confused -- isn't ```p``` still a ```BN_new()```-allocated object here?  I don't see ```p1``` getting freed, so it looks like it's leaking...

Comment (mspncp) [reply test/dhtest.c]: If I understand `DH_set0_pqg()` correctly, the three variables p, q, and g are assigned to members of the `dh` object without incrementing the reference count, so they are now owned by the dh object.

Comment (kaduk) [reply test/dhtest.c]: Ah, yes, that would be the piece I was missing -- thanks!
("set0" as in "0 additional references/copies are made")

File: test/dhtest.c (hunk 2 above)
Comment (mspncp) [reply test/dhtest.c]: I don't think we have a rule that forbids comma expressions ;-). But if anyone dislikes them, I will change it. I used it here to emphasize that the value moves from the one variable to the other.

File: test/dhtest.c (hunk 3 above)
Comment (romen) [reply test/dhtest.c]: Yes, it's like @mspncp says, but as @kaduk pointed, its' a bit confusing!

I suggest (https://github.com/openssl/openssl/pull/7155#discussion_r216167089) to add a comment to clarify (unless we go with the alternative solution https://github.com/openssl/openssl/pull/7155#discussion_r216167448)

File: test/dhtest.c (hunk 1 above)
Comment (mspncp) [reply test/dhtest.c]: Thanks for the suggestions. I'll think about them and push an update it as soon as possible.

Comment (romen) [reply test/dhtest.c]: Even better, if we moved lines 83 and 84 so they are part of the if starting at line 59, we could limit the `early_err` to only the first two `goto err`, avoiding a few corner cases.

What I am proposing basically wold be to group together:

```c
if (!TEST_true(BN_set_word(p, 4079L))
         || !TEST_true(BN_set_word(q, 2039L))
         || !TEST_true(BN_set_word(g, 3L))
         || !TEST_true(BN_set_word(priv_key, 1234L))
         || !TEST_true(DH_set0_pqg(dh, p, q, g))
         || !TEST_true(DH_set0_key(dh, NULL, priv_key)))
    goto early_err;
```

File: test/dhtest.c (hunk 3 above)
Comment (mspncp) [reply test/dhtest.c]: @romen I‘m tending towards your alternative proposal.

File: test/dhtest.c (hunk 1 above)
Comment (mspncp) [reply test/dhtest.c]: Well, not quite. But I get your point. There are three error cases, see the new fixup I pushed.

File: test/dhtest.c [hunk 4]
```diff
@@ -198,7 +208,6 @@ static int dh_test(void)
     BN_free(cpriv_key);
     BN_GENCB_free(_cb);
     DH_free(dh);
-
```
Comment (mspncp) [inline test/dhtest.c]: I added another fixup to restore this empty line which has been deleted unnecessarily. But I didn't push it yet in order not to interrupt the ci.

**Comment by mattcaswell** [comment]
Pushed.
